File: 130-surrounded-regions/surrounded-regions.cpp
class Solution {
public:
    pair<int, int> p[4] = {{0, -1}, {-1, 0}, {0, 1}, {1, 0}};
    void flooding(vector<vector<char>>& board, int i, int j)
    {
        board[i][j] = 'V';
        for (int k = 0; k < 4; k++)
        {
            int x = i + p[k].first;
            int y = j + p[k].second;
            if (x < board.size() && x >= 0 && y < board[0].size() && y >= 0 && board[x][y] == 'O')
            {
                flooding(board, x, y);
            }
        }
    }
    void solve(vector<vector<char>>& board) {
        int n = board.size();
        int m = board[0].size();

        for (int i = 0; i < n; i++)
        {
            if (board[i][0] == 'O') flooding(board, i, 0);
        }
        for (int i = 0; i < n; i++)
        {
            if (board[i][m - 1] == 'O') flooding(board, i, m - 1);
        }
        for (int j = 0; j < m; j++)
        {
            if (board[0][j] == 'O') flooding(board, 0, j);
        }
        for (int j = 0; j < m; j++)
        {
            if (board[n - 1][j] == 'O') flooding(board, n - 1, j);
        }
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < m; j++)
            {
                if (board[i][j] == 'V') board[i][j] = 'O';
                else if (board[i][j] == 'O') board[i][j] = 'X';
            }
        }
        
    }
};

File: 130-surrounded-regions/surrounded-regions.java
class Solution {
    public void solve(char[][] board) {
        int m = board.length;
        int n = board[0].length;

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (board[i][j] == 'X') continue;
                if (board[i][j] == 'O' && isEdge(i, j, m, n)) {
                    dfs(i, j, m, n, board);
                }
            }
        }

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (board[i][j] == 'O') board[i][j] = 'X';
            }
        }

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (board[i][j] == '#') board[i][j] = 'O';
            }
        }
    }

    private void dfs(int r, int c, int m, int n, char[][] board) {
        board[r][c] = '#';
        int[][] dirs = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};

        for (int[] dir : dirs) {
            int nr = r + dir[0];
            int nc = c + dir[1];

            if (isNotInside(nr, nc, m, n)) continue;
            if (board[nr][nc] == 'X' || board[nr][nc] == '#') continue;
            
            dfs(nr, nc, m, n, board);
        }
    }

    private boolean isEdge(int r, int c, int m, int n) {
        return r == 0 || c == 0 || r == m - 1 || c == n - 1;
    }

    private boolean isNotInside(int r, int c, int m, int n) {
        return r < 0 || c < 0 || r >= m || c >= n;
    }
}